Python/bltinmodule.c

```diff
@@ -2551,14 +2574,41 @@ zip_next(zipobject *lz)
 }
 
 static PyObject *
+zip_len(zipobject *lz)
+{
+    Py_ssize_t tuplesize = lz->tuplesize;
+    Py_ssize_t length_hint = PY_SSIZE_T_MAX;
+    Py_ssize_t i;
+
+    if (tuplesize == 0) {
+        Py_INCREF(_PyLong_Zero);
+        return _PyLong_Zero;
+    }
+
+    for (i = 0; length_hint > 0 && i < PyTuple_GET_SIZE(lz->ittuple); i++) {
+        PyObject *it = PyTuple_GET_ITEM(lz->ittuple, i);
+        Py_ssize_t it_len = PyObject_LengthHint(it, 0);
+
+        if (it_len < length_hint)
+            length_hint = it_len;
+    }
+
+    if (length_hint == -1)
+        return NULL;
+
```

Comment:
This looks incorrect to me.  Shouldn't it test if `it_len == PY_SSIZE_T_MAX`?  If this is a bug, your unit tests are not extensive enough.  Add tests that feed generators to zip and map and check to ensure that `__length_hint__` returns the correct value when no length can be determined.